Add JIT-EE interface wrappers to handle exceptions

JohnChen0 opened the pull request:
Added a managed wrapper to catch managed exceptions thrown by the JIT-EE
interface implementation. Added an unmanaged wrapper to re-throw the
exceptions for the compiler.

TBD: Capture message and call stack of the managed exception, and define
a custom unmanaged exception to include these information.

dnfclas commented:
Hi __@JohnChen0__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

jkotas commented on src/JitInterface/src/CorInfoImpl.cs:
```diff
@@ -41,6 +44,13 @@ public CorInfoImpl(Compilation compilation)
 
             _comp = CreateUnmanagedInstance();
 
+            if (!IsWindows())
```
We should do this everywhere - on both Windows and Unix.

jkotas replied:
(Note that we eventually want to self-host this toolchain on CoreRT, but we do not plan to build the managed-native exception for CoreRT - even on Windows.)

JohnChen0 replied:
Working on it.

jkotas commented:
Could you please remove _skipJitList from Compilation.cs and make sure that everything still works?

jkotas commented:
cc @kyulee1 @russellhadley It would be nice to push the unmanaged part of this wrapper into the JIT (ie build it as part of the JIT .dll). What do you think? 

Among other things, it would make the life easier for llilc - it would not need to depend on the private CoreCLR exception macros and EH implementation details.

kyulee1 commented:
@jkotas Pushing the unmanaged part of this wrapper into the JIT makes sense instead of having additional DLL for this purpose. The wrapper layout seems to be auto-generated/flattened from ICorJit data structure. Does it mean we need such auto-generation step (TunkGeneration) integreated inside RyuJit and Llilc for each? Is this wrapper (if integrated inside Jit) only needed for CoreRT where EE is managed?

jkotas commented:
> Does it mean we need such auto-generation step (TunkGeneration) integrated inside RyuJit and Llilc for each?

The step to auto-generate the wrapper is manual right now. I do not think it is worth the effort to make it automatic (to be done as part of every build). Maybe the wrapper can be part of the JIT SDK.

> Is this wrapper (if integrated inside Jit) only needed for CoreRT where EE is managed?

Yes. I am thinking that RyuJIT .dll can have two exports: getJit - that does exactly what it does today; and getJit2 - that goes through the wrapper.

gkhanna79 commented:
@jkotas Is there a specific reason to have the two exports instead of just one that allows the exceptions pass through since we would always need to do that?

jkotas commented:
The same RyuJIT .dll can be used by either CoreCLR or ilc today. I think it would be really nice to preserve this capability for engineering efficiency (assuming it does not cost too much, also there are different opinions about this across the JIT team).

schellap commented:
@kyulee1, I think you could also use this shim to hook up LLILC behind the scenes from corert, based on some env setting.

JohnChen0 commented:
@jkotas @MichalStrehovsky Could you please take a look?

Updated this PR to resolve known issues. The following changes are present:
* When JIT calls into the JIT-EE interface, the calls is wrapper. All managed exceptions are caught by the wrapper, and re-throw as native exceptions.
* The call from ILCompiler to JIT is also wrapper. The wrapper catches all native exceptions, and re-throws them as managed exceptions.
* The message and call stack from the original managed exception is captured and stored in the unmanaged exception object, and restored in the managed exception thrown by the JIT wrapper.
* A new native module (jitinterface.dll or jitinterface.so) is added to contain the native part of the wrappers. Long term, we should consider whether to merge this into JIT.
* Removed most, but not all, entries from _skipJitList. A couple of methods still cause errors unrelated to exception handling.

jkotas commented:
> Removed most, but not all, entries from _skipJitList. A couple of methods still cause errors unrelated to exception handling

What kind of errors are these? They should be caught by the toolchain.

JohnChen0 commented:
They are undefined reference errors. Here are the error messages. Only Hello.dll fails. The other two succeeds.

Input Assembly: /home/john/corert/tests/src/Simple/Hello/bin/Debug/dnxcore50/Hello.dll
/home/john/corert/tests/src/Simple/Hello/obj/Debug/dnxcore50/Hello.obj: In function `jmpstub_GetNativeSystemInfo':
(.text+0x23975): undefined reference to `GetNativeSystemInfo'
/home/john/corert/tests/src/Simple/Hello/obj/Debug/dnxcore50/Hello.obj: In function `jmpstub_CreateEventExW':
(.text+0x280f1): undefined reference to `CreateEventExW'
/home/john/corert/tests/src/Simple/Hello/obj/Debug/dnxcore50/Hello.obj: In function `jmpstub_CoGetApartmentType':
(.text+0x28a55): undefined reference to `CoGetApartmentType'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Compilation of intermediate files failed.

MichalStrehovsky commented:
@JohnChen0 You're compiling against the Windows version of CoreLib. You need use the one you built on Linux (or pass /p:TargetsLinux=true to msbuild I *think*).

jkotas commented:
We have hit portion of System.CoreLib that was not ported to Unix yet. We should add a stubs for these methods to src\Native\Bootstrap\main.cpp and open issues to get it properly ported.

jkotas commented on src/JitInterface/src/ThunkGenerator/Program.cs:
```diff
@@ -272,6 +273,88 @@ unsafe partial class CorInfoImpl
+            }
+            catch (Exception ex)
+            {
+                string exString = ex.ToString();
```
We should hoist these two lines into a helper method - to reduce code bloat.

JohnChen0 replied:
Done

jkotas commented on src/Native/jitinterface/jitwrapper.cpp:
```diff
@@ -0,0 +1,44 @@
+    {
+        *ppException = pException;
+    }
+    catch (...)
```
It this needed for anything? We should not be catching somebody else's exceptions.

gkhanna79 replied:
I concur. Infact, this will endup catching the OOM from AllocException (assuming JIT didnt catch it).

JohnChen0 replied:
Removed catch (...) from the wrapper

jkotas commented on src/Native/jitinterface/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+
+if (WIN32)
+    list(APPEND NATIVE_SOURCES jitinterface.def)
+    add_compile_options(/Zi)
```
This should be in the global CMakeFile one level up (same as in CoreCLR). We want to have debug symbols for every library that we build.

JohnChen0 replied:
Moved PDB related settings from Native/jitinterface/CMakeLists.txt to Native/CMakeLists.txt.

jkotas commented on src/JitInterface/src/CorInfoImpl.cs:
```diff
@@ -2145,12 +2165,33 @@ private ushort getRelocTypeHint(IntPtr _this, void* target)
+            {
+                IntPtr * vtable = *(IntPtr**)_comp;
+                for (int i = 0; i < _keepalive.Length; i++)
+                {
```
We should add concept of manually written C++  wrapper. It should be used for getLocationOfThisType and FilterException:

- getLocationOfThisType: the managed side should always get the return value as [out] argument. The workaround on the managed side should go away, and the  special cases on the unmanaged side will be centralized.

- FilterException should just unconditionally return EXECUTE_HANDLER on the C++ side without calling managed code. (It is a reliability problem to write Win32 unmanaged exception filters in managed code.)

JohnChen0 replied:
Added the ability to manually write C++ wrappers. Auto-generated wrappers are in jitinterface.h, while manually written ones are in jitinterface.cpp.

jkotas commented on src/JitInterface/src/ThunkGenerator/Program.cs:
```diff
@@ -272,6 +273,88 @@ unsafe partial class CorInfoImpl
+                            tr.Write("null");
+                            break;
+
+                        case "CORINFO_LOOKUP_KIND":
```
Should not be needed with manual getLocationOfThisType wrapper.

JohnChen0 replied:
Removed

jkotas commented on src/JitInterface/src/ThunkGenerator/Program.cs:
```diff
@@ -302,6 +385,102 @@ protected IntPtr CreateUnmanagedInstance()
+#include ""corinfoexception.h""
+
+enum CORINFO_RUNTIME_LOOKUP_KIND { };
+struct CORINFO_LOOKUP_KIND
```
Should not be needed with manual getLocationOfThisType wrapper.

JohnChen0 replied:
Removed this from the auto-generated code.

jkotas commented on src/JitInterface/src/CorInfoImpl.cs:
```diff
@@ -94,9 +106,17 @@ public void CompileMethod(MethodCodeNode methodCodeNodeNeedingCode)
+                {
+                    char* szMessage = GetExceptionMessage(exception);
+                    string message = szMessage != null ? new string(szMessage) : "JIT Exception";
+                    FreeException(exception);
```
The exception can be caught by the JIT too - we will leak memory if the JIT catches the exception (e.g. if the exception happens during inlining - it is not fatal). To avoid the leak, we may want to keep track of all exceptions that were thrown during JITing and free them in FlushPins (this method can use a better name...).

JohnChen0 replied:
Added a linked list to keep track of all exceptions, and free them in FlushPins (renamed to CompileMethodCleanup).

jkotas commented:
LGTM otherwise. Thanks!

gkhanna79 commented on src/JitInterface/src/ThunkGenerator/Program.cs:
```diff
@@ -272,6 +273,88 @@ unsafe partial class CorInfoImpl
+            foreach (FunctionDecl decl in functionData)
+            {
+                string returnType = decl.ReturnType.ManagedTypeName;
+                int marshalAs = returnType.LastIndexOf(']');
```
Why is this called "marshalAs"? Is this checking presence of MarshalAs attribute associated with return type? If so, will this handle return type that are arrays (e.g. string[])?

JohnChen0 replied:
Only MarshalAs attributes are expected to be used inside the square brackets. Not sure if we can handle string[], though this is not intended to be a general-purpose thunk generator. It only needs to handle types used in JIT-EE interface.

JohnChen0 commented:
Updated the PR to address all review feedbacks. _skipJitList is now completely gone.

jkotas commented on src/Native/jitinterface/corinfoexception.cpp:
```diff
@@ -0,0 +1,50 @@
+    CorInfoExceptionList* _pNext;
+};
+
+static CorInfoExceptionList *pExceptionList = nullptr;
```
The wrapper should be multi-threading and recursion friendly. This list should be per compilation instance - I think that the easiest way to make it happen is to maintain it at the managed side.

JohnChen0 replied:
Fixed by tracking the exceptions in a List in managed code. Insertion into the List is protected by lock to be multi-threading safe.

jkotas commented on src/JitInterface/src/CorInfoImpl.cs:
```diff
@@ -1649,38 +1672,30 @@ private bool IsWindows()
             return Path.DirectorySeparatorChar == '\\';
```
The `IsWindows()` method should be deleted - it was part of workaround that you have removed.

JohnChen0 replied:
Removed

jkotas commented on src/JitInterface/src/CorInfoImpl.cs:
```diff
@@ -1373,26 +1396,26 @@ private CorInfoTypeWithMod getArgType(IntPtr _this, CORINFO_SIG_INFO* sig, CORIN
-        private int FilterException(IntPtr _this, _EXCEPTION_POINTERS* pExceptionPointers)
+        private int FilterException(_EXCEPTION_POINTERS* pExceptionPointers)
         {
             return 0; // EXCEPTION_CONTINUE_SEARCH
```
It would be nice to comment, assert and throw to make it clear that this method is not expected to be called because of it is implemented on the unmanaged side

JohnChen0 replied:
Done

jkotas commented on src/JitInterface/src/ThunkGenerator/Program.cs:
```diff
@@ -302,6 +411,103 @@ protected IntPtr CreateUnmanagedInstance()
+#include <stdlib.h>
+#include ""corinfoexception.h""
+
+struct CORINFO_LOOKUP_KIND;
```
Still needed?

JohnChen0 replied:
A basic declaration of the type still appears needed, since I need to include the prototype of getLocationOfThisType method in JitInterfaceWrapper to take its space in vtable, with call signature matching what JIT expects.

jkotas commented:
LGTM module comments

jkotas commented on src/JitInterface/src/CorInfoImpl.cs:
```diff
@@ -150,12 +176,19 @@ private IntPtr GetPin(Object obj)
         }
-        private void FlushPins()
+
+        private List<IntPtr> _nativeExceptions = new List<IntPtr>();
```
This list should be created lazily - only once there is exception to store into it.

JohnChen0 replied:
Updated the PR to create this list lazily, to remove lock around modifying it, and to set the list to null after using it.

jkotas commented on src/JitInterface/src/CorInfoImpl.cs:
```diff
@@ -26,13 +26,34 @@ internal unsafe partial class CorInfoImpl
+        {
+            string exString = ex.ToString();
+            IntPtr nativeException = AllocException(exString, exString.Length);
+            lock (_nativeExceptions)
```
This lock is not necessary - we assume in many places that the given CorInfoImpl instance is used from a single thread only.

jkotas commented on src/JitInterface/src/CorInfoImpl.cs:
```diff
@@ -150,12 +176,19 @@ private IntPtr GetPin(Object obj)
 
+            foreach (IntPtr ex in _nativeExceptions)
+                FreeException(ex);
+            _nativeExceptions.Clear();
```
`nativeExceptions = null` would be better instead of Clear - once you start creating the list lazily.

jkotas commented:
:+1:
